Sources/SwiftProtobuf/PopulatorConstructable.swift

```diff
@@ -0,0 +1,46 @@
+// Sources/SwiftProtobuf/PopulatorConstructable.swift
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+
+
+import Foundation
+
+public protocol PopulatorConstructable {
```

Comment:
I'll defer to others if it's worth doing this as a protocol, but the general idea does seem like it might make sense to support it on all of our `*Options` types (either via a protocol or by just manually adding it.

Reply:
My own opinion:

I'm personally not a huge fan of ad hoc APIs like this (as the person who introduced the `with` helper for `Message`, I regret it today). There's been some recent discussion on the Swift forums about solving this in a more language-native way, and while this isn't something that's on the near horizon, I'd still prefer to not introduce another API that might have a better solution later on; having to declare a new `var` isn't _terribly_ onerous.

In general, when introducing a new protocol, a good question to ask is "what algorithms would I write generically over types that have that conformance requirement?" In this case, I think the answer is "none" aside from the `with` API. Providing this protocol tempts users to conform to it themselves or write generic code that uses it, which isn't something we want. In fact, I think it might be the case that a user seeing `.with` used in protobuf code would make it very easy for them to think the options types were _themselves_ protobufs by association, because of the existing inertia around that method.

Reply:
I agree with Tony about this protocol:  It doesn't really serve any purpose since there's little point in trying to build generic or existential code around it.  (The closure has to know exactly what fields the type supports, so the closure cannot itself be generic.)  Since every protocol conformance generates runtime metadata specifically to support existential or generic code, we should try to avoid protocols that aren't necessary. That's independent of whether `.with()` is a good API for these option structs.

Regarding `.with()`:   We added `.with()` to the message API in part because the obvious alternative -- using initializers with default arguments -- does not scale well to objects with dozens or hundreds of optional fields.  That's not an issue for these option structs.  The other motivating concern was around the appropriate ordering of such arguments -- again, that's not an issue for these option structs.  We can easily create static values of the option structs and/or initializers with default arguments and/or multiple initializers for common use cases.

(I'm sympathetic to Tony's argument that creating a `var` isn't that much friction, though I'm not entirely convinced I agree. :grin:)

Reply:
I still think theres value in being able to declare an immutable `*Options` instance in a single statement without having to resort to a closure. It doesn't really matter to me whether thats done through an initializer with default values or through a `with` constructor but I do think its valuable nonetheless. I think its easier to parse, and I think it provides safety guaranteeing that the value will not be accidentally mutated elsewhere in the function. 

The protocol was just an attempt to consolidate the logic but I'm happy to just duplicate the `with` implementation across different`*Options` or to create a non-empty initializer with default values.

Reply:
Removed the protocol for the sake of this PR